### PR TITLE
Add Support for Parameter Binding Arrays

### PR DESCRIPTION
**Pull request by athmihir**
*Issue number of the reported bug or feature request: #<number>*

**Describe your changes**
Add the ability to parameter bind carrays in dbapi2 through python lists, or tuples of type: int, float. string or bytes.
This is by exposing the cdb2_bind_array function present in cdb2api.

**Testing performed**
Add pytests that try all combinations of accepted parameters. ie. lists and tuples of allowed data types.

**Additional context**
None.

### CODE REVIEW

**Comment by godlygeek** [comment]
I've taken a first look at this. It looks quite good, though I've made some small improvements. I plan to take another look tomorrow.

**Comment by athmihir** [comment]
@godlygeek sounds good. Looks like DCO is failing

**Comment by godlygeek** [comment]
> Looks like DCO is failing

Yep, that's fine. I didn't bother signing my fixup commits, since I plan to squash them into your commits.

**Comment by godlygeek** [comment]
This PR now looks good to me, though I'd like another set of eyes on it (probably @gusmonod) since I made some changes here myself.

We're going to hold off on merging this PR for a little while, though. We're gonna merge this after we cut a new release that officially drops support for versions prior to Python 3.7.

**Comment by athmihir** [comment]
Sounds good, thanks for the review. Looking forward to getting it merged! 😄

File: comdb2/_ccdb2.pyx
Comment (gusmonod) [inline comdb2/_ccdb2.pyx:188]: Why call `PyMem_Malloc` here? Aren't we using twice as much memory as we need to?

I think `self.owner` could be initialized to `obj`, which would increase the reference count by one and ensure it doesn't get garbage collected until nothing has a reference to the `_ParameterValue` instance, and we can just make `self.data` point to the data held by `self.owner`.

File: comdb2/_ccdb2.pyx
Comment (gusmonod) [inline comdb2/_ccdb2.pyx:198]: What about here, why call `PyMem_Malloc`? I know that the `__dealloc__` was changed so that it calls `PyMem_Free` every time that `self.list_size != -1`, so there is no leak, but aren't we using twice as much memory as we need to once again?

In this case I think `self.owner` is initialized with a list that is a new copy with all of the data we need there, in a Python object that has its reference count set properly, so we can just make `self.data` point to the data held by `self.owner`. Am I missing something?

File: tests/test_dbapi2.py
Comment (gusmonod) [inline tests/test_dbapi2.py:907]: I would add another test case for nested lists or nested tuples, just to check that the error message makes sense for them too.

**Review by gusmonod — CHANGES_REQUESTED** [review]
There is one thing that doesn't make sense to me, and that is the allocation of the memory for the arrays of bytes.

I wonder if maybe it needs to be done because of the fact that the data in python `list`s and `tuple`s may not be contiguous… 🤔 either way, would like to know what you all think about it.

File: comdb2/_ccdb2.pyx
Comment (godlygeek) [reply comdb2/_ccdb2.pyx:188]: > Why call `PyMem_Malloc` here? Aren't we using twice as much memory as we need to?

We need to pass a contiguous array of `size_t` / `char*` pairs to libcdb2api, and that's not something that we can export from the CPython C list/tuple API - we need to allocate memory for that array, and free it when we're done.

> I think `self.owner` could be initialized to `obj`, which would increase the reference count by one and ensure it doesn't get garbage collected until nothing has a reference to the `_ParameterValue` instance

I was going to say that we don't need to worry about this - that `parameters` itself outlives the `_ParameterValue` instances, and would keep `obj` alive. But, that's not quite true: while we're making the call to `cdb2_run_statement` we drop the GIL, and once we drop the GIL another thread could mutate the `parameters` dict so that it no longer references `obj`. So, I think you're half right - we _do_ need to set `self.owner = obj` here, even though we _also_ need to malloc a contiguous array.

File: tests/test_dbapi2.py
Comment (godlygeek) [reply tests/test_dbapi2.py:907]: It'll just say "Cannot bind a tuple of tuple" (since we only go one level deep), but sure, that's an easy test to add.

File: comdb2/_ccdb2.pyx
Comment (athmihir) [reply comdb2/_ccdb2.pyx:188]: Yeah, I guess it got overridden during Matt's changes. Adding it back.

File: comdb2/_ccdb2.pyx
Comment (athmihir) [reply comdb2/_ccdb2.pyx:198]: So the python object holds a list of strings. This is allocated on the heap by python during runtime.
For example a = ['apple', 'banana', 'orange']

The pyMem_Malloc creates a C array of pointers. The pointers themselves do not store the entire strings but just the starting memory locations. Each pointer is only 64 bits long even though the underlying string it points to could be 
as long as python wants. So, we aren't allocating memory "twice".
We need to pass an array of pointers or an array of "char*" so that the c api knows where each string starts from.
Then it just reads from the memory allocated in the python list.

How does it know when to stop? This is why we need to do x.encode('utf-8'). This ensures the string terminates
with a null character. something that is common in C-style strings.

For ints and floats we already know they are of fixed length so allocate memory directly for longs, or ints.
Same cannot be said about a strings or bytes hence the usage of pointers.

Basically a char** is an array of char* where each char* is the first byte for each string in our python array.

**Pull request merged**
